### PR TITLE
Fix torch.where to accept only tensors with same dtypes(CPU)

### PR DESCRIPTION
**Pull request by ifedan**
Currently torch.where can be called with tensor arguments of different types, as an example: #28709

This fix should prohibit this and allow to call torch.where only with argument of same type. It's similar to CUDA behaviour.

### CODE REVIEW

**Comment by ifedan** [comment]
Same change for master: https://github.com/pytorch/pytorch/pull/29081

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ Tensor cosine_embedding_loss(const Tensor& input1, const Tensor& input2, const T
   auto denom = (mag_square1 * mag_square2).sqrt_();
   auto cos = prod_sum / denom;
 
-  auto zeros = at::zeros_like(target);
+  auto zeros = at::zeros_like(cos);
```
Comment (nairbv) [inline aten/src/ATen/native/Loss.cpp]: can both inputs be integral? margin is a double so that would make neg a double (below), used in output_neg

Comment (ifedan) [reply aten/src/ATen/native/Loss.cpp]: Yes, I think the way we should go with cosine_embedding_loss and hinge_embedding_loss is something like:
```
auto zeros = at::zeros_like(pos);
auto output_pos = at::where(target == 1, pos, zeros);
auto output_neg = at::where(target == -1, neg, zeros.**to(neg.dtype())**);
```

Comment (gchanan) [reply aten/src/ATen/native/Loss.cpp]: I don't see how the inputs being integral matter:
```
auto mag_square1 = (input1 * input1).sum(1) + EPSILON;
auto mag_square2 = (input2 * input2).sum(1) + EPSILON;
```

this line will make everything floating point anyway.

Of course, you should just test this.

File: test/test_torch.py [hunk 2]
```diff
@@ -951,6 +951,45 @@ def test_where_bool_tensor(self):
             res = torch.where(a > 0)
             self.assertEqual(1, len(res))
 
+    def test_where_tensor_cpu(self):
+        def rand_tensor(size, dtype, device, cont):
```
Comment (gchanan) [inline test/test_torch.py]: cont isn't used in this function.

File: test/test_torch.py [hunk 3]
```diff
@@ -951,6 +951,45 @@ def test_where_bool_tensor(self):
...
+                return torch.rand(size=size, dtype=dtype, device=device)
+            elif dtype == torch.uint8:
+                return torch.randint(1, 5, size=size, dtype=dtype, device=device)
+            elif dtype == torch.bool:
+                return torch.randint(0, 1, size=size, dtype=dtype, device=device).bool()
+            else:
+                return torch.randint(-5, 5, size=size, dtype=dtype, device=device)
+
+        def get_tensor(size, dtype, device, cont):
+            if not cont and len(size) < 2:
```
Comment (gchanan) [inline test/test_torch.py]: can you use "contiguous" or "contig"?  There are too many words that could be relevant here...continuous, continue, etc.

File: test/test_torch.py [hunk 4]
```diff
@@ -951,6 +951,45 @@ def test_where_bool_tensor(self):
...
+        for device in torch.testing.get_all_device_types():
+            for dt1 in torch.testing.get_all_math_dtypes(device):
+                for dt2 in torch.testing.get_all_math_dtypes(device):
+                    for cont in [True, False]:
+                        x1 = get_tensor((height, width), dt1, device, cont)
+                        x2 = get_tensor((height, width), dt2, device, cont)
+                        if dt1 != dt2:
+                            self.assertRaisesRegex(RuntimeError, "expected scalar type", lambda: torch.where(x1 == 1, x1, x2))
+                        else:
+                            res = torch.where(x1 == 1, x1, x2)
```
Comment (gchanan) [inline test/test_torch.py]: wait -- is this going to very rarely happen?  Like, never if x1 is floating point?

**Review by gchanan — COMMENTED** [review]
how about a cosine_embedding_loss and kl_div test?

File: test/test_torch.py (hunk 2 above)
Comment (ifedan) [reply test/test_torch.py]: changed

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (ifedan) [reply aten/src/ATen/native/Loss.cpp]: Added test

File: test/test_torch.py (hunk 3 above)
Comment (ifedan) [reply test/test_torch.py]: changed

File: test/test_torch.py (hunk 4 above)
Comment (ifedan) [reply test/test_torch.py]: > how about a cosine_embedding_loss and kl_div test?

Added

**Comment by ifedan** [comment]
> how about a cosine_embedding_loss and kl_div test?

Added

File: aten/src/ATen/native/TensorCompare.cpp [hunk 5]
```diff
@@ -122,6 +122,9 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where_cpu(const Tensor& condition, const Tensor& self, const Tensor& other) {
+  if (self.dtype() != other.dtype()) {
```
Comment (gchanan) [inline aten/src/ATen/native/TensorCompare.cpp]: nit: it's better to just write `TORCH_CHECK(self.dtype() == other.dtype(), ...`.

File: test/test_torch.py [hunk 6]
```diff
@@ -951,6 +951,49 @@ def test_where_bool_tensor(self):
...
+                        if dt1 != dt2:
+                            self.assertRaisesRegex(RuntimeError, "expected scalar type", lambda: torch.where(x1 == 1, x1, x2))
+                        else:
+                            if x1.is_floating_point():
+                                res = torch.where(x1 < 0.5, x1, x2)
+                            else:
+                                res = torch.where(x1 == 1, x1, x2)
+                            for i in range(height):
+                                for j in range(width):
+                                    if (x1.is_floating_point() and x1[i][j] < 0.5) or \
```
Comment (gchanan) [inline test/test_torch.py]: you should just cache the results of the comparison so you don't have to reimplement this logic.

File: test/test_nn.py [hunk 7]
```diff
@@ -7223,6 +7223,30 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
                          loss_reference_fns['CosineEmbeddingLoss'](input1, input2, target,
                                                                    margin=0.5, reduction='none'))
 
+    def test_cosine_embedding_loss_with_diff_type(self):
+        for device in device_():
+            for dt1 in torch.testing.get_all_math_dtypes(device):
+                for dt2 in torch.testing.get_all_math_dtypes(device):
+                    for dt3 in torch.testing.get_all_math_dtypes(device):
+                        if dt3 == torch.uint8:
```
Comment (gchanan) [inline test/test_nn.py]: what happens here? (comment).

File: test/test_nn.py [hunk 8]
```diff
@@ -7223,6 +7223,30 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
...
+                            continue
+                        input1 = torch.tensor([[2, 3, 4], [6, 2, 4]], dtype=dt1, device=device)
+                        input2 = torch.tensor([[2, 3, 5], [3, 2, 1]], dtype=dt2, device=device)
+                        target = torch.tensor([1, -1], dtype=dt3, device=device)
+                        result = torch.nn.functional.cosine_embedding_loss(input1, input2, target)
+                        self.assertEqual(result.item(), 0.4672, 0.001)
```
Comment (gchanan) [inline test/test_nn.py]: instead of a magic number, why don't you compare with the case where everything is a double.

File: test/test_nn.py [hunk 9]
```diff
@@ -7223,6 +7223,30 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
...
+
+    def test_kl_div_with_diff_type(self):
+        for device in device_():
+            for input_dtype in torch.testing.get_all_math_dtypes(device):
+                for target_dtype in [torch.float32, torch.float64, torch.float16]:
+                    if not device.startswith('cuda') and target_dtype == torch.float16:
+                        continue
```
Comment (gchanan) [inline test/test_nn.py]: this is bad for XLA (which is implemented like cuda), do `not `cpu`, also use `torch.device(device).type` instead of startswith.

File: test/test_nn.py [hunk 10]
```diff
@@ -7223,6 +7223,30 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
...
+                    input = torch.tensor([[2, 3, 5], [3, 2, 1]], dtype=input_dtype, device=device)
+                    target = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=target_dtype, device=device)
+                    result = torch.nn.functional.kl_div(input, target)
+                    self.assertEqual(result.item(), -3.6625, 0.001)
```
Comment (gchanan) [inline test/test_nn.py]: again, compare with the all-double result.

File: test/test_nn.py [hunk 11]
```diff
@@ -7223,6 +7223,30 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
                          loss_reference_fns['CosineEmbeddingLoss'](input1, input2, target,
                                                                    margin=0.5, reduction='none'))
 
+    def test_cosine_embedding_loss_with_diff_type(self):
+        for device in device_():
+            for dt1 in torch.testing.get_all_math_dtypes(device):
```
Comment (gchanan) [inline test/test_nn.py]: how long does this test take?  It's like...1000 combinations?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/TensorCompare.cpp (hunk 5 above)
Comment (ifedan) [reply aten/src/ATen/native/TensorCompare.cpp]: Changed

File: test/test_nn.py (hunk 7 above)
Comment (ifedan) [reply test/test_nn.py]: Added a comment

File: test/test_nn.py (hunk 8 above)
Comment (ifedan) [reply test/test_nn.py]: Changed

File: test/test_nn.py (hunk 9 above)
Comment (ifedan) [reply test/test_nn.py]: Changed

File: test/test_nn.py (hunk 10 above)
Comment (ifedan) [reply test/test_nn.py]: Changed

File: test/test_nn.py (hunk 11 above)
Comment (ifedan) [reply test/test_nn.py]: It takes about 1 sec

File: test/test_torch.py (hunk 6 above)
Comment (ifedan) [reply test/test_torch.py]: Changed to:
```
expected = condition.to(x1.dtype) * x1 + (~condition).to(x2.dtype) * x2
result = torch.where(condition, x1, x2)
self.assertEqual(expected, result)
```

File: test/test_nn.py (hunk 11 above)
Comment (ifedan) [reply test/test_nn.py]: Update: it takes around 300ms

File: test/test_nn.py [hunk 12]
```diff
@@ -7223,6 +7223,38 @@ def test_cosine_embedding_loss_margin_no_reduce(self):
                          loss_reference_fns['CosineEmbeddingLoss'](input1, input2, target,
                                                                    margin=0.5, reduction='none'))
 
+    def test_cosine_embedding_loss_with_diff_type(self):
+        for device in device_():
+            input1 = torch.tensor([[2, 3, 4], [6, 2, 4]], dtype=torch.double, device=device)
+            input2 = torch.tensor([[2, 3, 5], [3, 2, 1]], dtype=torch.double, device=device)
+            target = torch.tensor([1, -1], dtype=torch.int, device=device)
```
Comment (gchanan) [inline test/test_nn.py]: nit: its more standard to use dtype long here.

**Review by gchanan — APPROVED** [review]
